### PR TITLE
Found an issue which causing the App crash.

### PR DESCRIPTION
**Pull request by inampaki**
Resolved an issue which crashing the application that when string is passed to the object in 2nd argument instead of Array

```javascript
sender.send(message, { registrationTokens: "device id"}, function (err, response) {
});
```

### CODE REVIEW

**Comment by eladnava** [comment]
@inampaki Thanks for your contribution!

I'm not sure that this is behavior that we'd like to support -- the documentation clearly demonstrates that the `sender.send` method accepts `{ registrationTokens: regTokens }` as an array, and the name of the property `registrationTokens` is plural which would indicate an array.

```js
var regTokens = ['YOUR_REG_TOKEN_HERE'];

// Now the sender can be used to send messages
sender.send(message, { registrationTokens: regTokens }, function (err, response) {
    if(err) console.error(err);
    else    console.log(response);
});
```

@hypesystem What do you think?

**Comment by inampaki** [comment]
I agree with you but in that case we should have some error returned instead of App crash due to undfined in response key at line 207

**Comment by eladnava** [comment]
@inampaki You're correct, it should definitely return an error to the callback provided and not crash the app.

Let's see if @hypesystem agrees, in that case, you can modify your PR to return an error instead of automatically converting the string to an array.

**Comment by inampaki** [comment]
@eladnava I have returned an error message as we discussed and reverted the conversion of object value to array as it was originally.

**Comment by eladnava** [comment]
@inampaki See my comment.

File: lib/sender.js [hunk 1]
```diff
@@ -29,18 +29,27 @@ var parseAndRespond = function(resBody, callback) {
 function extractRecipient(recipient) {
     var validKeys = ['registrationIds', 'registrationTokens', 'topic', 'notificationKey'];
     var theRecipient;
-
-    validKeys.forEach(function(key) {
-        if (recipient[key]) {
+    var err = null;
+    var key = null;
+    for(var i = 0; i < validKeys.length; i++){
```
Comment (hypesystem) [inline lib/sender.js]: I would like to avoid using native for-loops. forEach should be preferred.

**Comment by hypesystem** [comment]
First of all, thank you for your time, and well spotted error!

I agree that returning a legible error when passing a string where array is expected is a good idea.

This current code, as I see it, is less readable than the code before, and needs to be improved before merge. I will try to attach a few comments. In addition, @eladnava is correct in noting that it does not support topic and notificationKey properly. You would have seen this had you run `npm test`.

File: lib/sender.js [hunk 2]
```diff
@@ -29,18 +29,27 @@ var parseAndRespond = function(resBody, callback) {
...
+                break;
+            }
             theRecipient = recipient[key];
-            return;
+            break;
         }
-    });
+    }
+
+    if(i == validKeys.length){
```
Comment (hypesystem) [inline lib/sender.js]: You could just do `if(err)`

File: lib/sender.js [hunk 3]
```diff
@@ -29,18 +29,27 @@ var parseAndRespond = function(resBody, callback) {
 function extractRecipient(recipient) {
     var validKeys = ['registrationIds', 'registrationTokens', 'topic', 'notificationKey'];
     var theRecipient;
-
-    validKeys.forEach(function(key) {
-        if (recipient[key]) {
+    var err = null;
+    var key = null;
```
Comment (hypesystem) [inline lib/sender.js]: There is no reason to instantiate `key` here. It is never used outside the loop.

Comment (hypesystem) [inline lib/sender.js]: Also when instantiating, there's no reason to set the variables to null.

```js
var err; // <-- instantiates `err`, set to `undefined`
```

**Comment by hypesystem** [comment]
Now that we're at it, I think we could probably extract the whole `extractRecipient` function to its own file, as it is entirely standalone. This would make the sender **and** the extractRecipient code easier to get an overview of.

**Comment by inampaki** [comment]
@hypesystem thanks for the tips, you are right I think I should close this issue then.

File: lib/sender.js (hunk 2 above)
Comment (inampaki) [inline lib/sender.js]: in case of invalid key err will be null but i will be equal to the length of validKeys that's why I checked it like this.

**Pull request closed without merging**